dnf/transaction_sr.py

```diff
@@ -0,0 +1,359 @@
+                        continue
+
+                    self._base.group_install(group_id, pkg_types)
+                elif action == "Upgrade":
+                    self._base.group_upgrade(group_id)
+                elif action == "Removed":
+                    self._base.group_remove(group_id)
+                else:
+                    errors.append(TransactionError(
+                        _('Unexpected value of group action "{action}" for group "{group}".')
+                            .format(action=action, group=group_id)
+                    ))
+            except KeyError as e:
+                errors.append(TransactionError(
+                    _('Missing object key "{key}" in a group.').format(key=e.args[0])
+                ))
+            except ValueError as e:
+                errors.append(TransactionError(str(e)))
+            except dnf.exceptions.CompsError as e:
+                errors.append(TransactionError(str(e)))
+
+        for env_data in environments:
+            try:
+                action = env_data["action"]
+                env_id = env_data["id"]
+
+                if action == "Install":
+                    try:
+                        pkg_types = libdnf.transaction.stringToCompsPackageType(env_data["package_types"])
+                    except libdnf.error.Error as e:
+                        errors.append(TransactionError(str(e)))
+                        continue
+
+                    self._base.environment_install(env_id, pkg_types)
+                elif action == "Upgrade":
+                    self._base.environment_upgrade(env_id)
+                elif action == "Removed":
+                    self._base.environment_remove(env_id)
+                else:
+                    errors.append(TransactionError(
+                        _('Unexpected value of environment action "{action}" for environment "{env}".')
+                            .format(action=action, env=env_id)
+                    ))
+            except KeyError as e:
+                errors.append(TransactionError(
+                    _('Missing object key "{key}" in an environment.').format(key=e.args[0])
+                ))
+            except ValueError as e:
+                errors.append(TransactionError(str(e)))
+            except dnf.exceptions.CompsError as e:
+                errors.append(TransactionError(str(e)))
+
+        if errors:
+            raise TransactionFileError(fn, errors)
+
+    def post_transaction(self):
+        """
+        Sets reasons in the transaction history to values from the stored transaction.
+
+        Also serves to check whether additional packages were pulled in by the
+        transaction, which results in an error.
+        """
+        if not self._base.transaction:
+            return
+
+        errors = []
+
+        for tsi in self._base.transaction:
+            try:
+                tsi.reason = self._pkg_reason_cache[str(tsi.pkg)]
```

Comment:
During reinstall you will override both reinstall and reinstalled by the same reasons. That could be confusing. Also I am not sure how you want to handle combination of reason change and upgrade?

Reply:
Probably you can take look in system-upgrade how it handles reasons.

Reply:
This is given by the ambiguous way the transaction items are stored. Are the reasons actually different for the "Reinstall" and "Reinstalled" actions respectively? (Can you think of a test case?)

Having a look at the systme-upgrade plugin, it seems to be doing the exact same thing as my code... So it either has the same problem or it's actually not a problem since the reasons for the actions match? I don't know.

Reply:
I am sorry, my mistake, I should probably copy the code for you directly. The code from `system-upgrade` is not the same. See bellow. As you can see it stores reasons for forward actions and reverse actions in separate dicts. Also there is another approach how the information is applied. In your case you have to also consider reason-change action that is not present in system-upgrade.
```
        for tsi in self.cli.base.transaction:
            if tsi.action in forward_actions:
                pkg = tsi.pkg
                forward.setdefault(pkg.repo.id, {}).setdefault(
                    str(pkg), {})[tsi.action] = tsi.reason
            elif tsi.action in backward_action:
                reverse.setdefault(str(tsi.pkg), {})[tsi.action] = tsi.reason
```
Here is the code how system-upgrade applies reasons. Again it is not the same code like yours. As you can see it handles forward and reverse operation separately. Additional it is more flexible. When some mismatches are there it skips it.
```
install_packages = self.state.install_packages
        remove_packages = self.state.remove_packages
        for tsi in self.cli.base.transaction:
            if tsi.action in forward_actions:
                pkg = tsi.pkg
                try:
                    stored_reason = install_packages[pkg.repo.id][str(pkg)][str(tsi.action)]
                    if stored_reason != tsi.reason:
                        tsi.reason = stored_reason
                except KeyError:
                    pass
            elif tsi.action in backward_action:
                pkg = tsi.pkg
                try:
                    stored_reason = remove_packages[str(pkg)][str(tsi.action)]
                    if stored_reason != tsi.reason:
                        tsi.reason = stored_reason
                except KeyError:
                    pass
```

Reply:
I see. It stores the forward and backward actions separately so the reasons for each don't overwrite each other, as is the case with my code. How do you suggest to solve the issue though? I can't think of a solution that isn't ugly. Should I split the reason cache to `forward_reason_cache` and `backward_reason_cache`? Or just extend the key to differentiate? But with what?

Reply:
There are also additional difference
1. The code in system-upgrade uses action - reason pairs
2. Reasons are modified only when action fits with transaction item action. When actions differ it simply skip it.

Reply:
> Reasons are modified only when action fits with transaction item action. When actions differ it simply skip it.

I find this problematic, for example, if there is a `user install foo-2.0` in the transaction file and `foo-1.0` is already installed on the target system as a dependency, it will end up being an upgrade `foo-1.0 -> foo-2.0` but it will not be marked as userinstalled by your logic. Which it certainly should?